### PR TITLE
executor: start hash join inner/outer workers together

### PR DESCRIPTION
**Pull request by zz-jason**
## What have you changed? (mandatory)

Before this PR, the execution flow of the hash join operator is:
1. fetch inner rows and build hash table based on the fetched inner rows
2. use a goroutine(`fetchOuterChunks()`) to fetch outer rows and dispatch the fetched outer rows to many hash join workers(`runJoinWorker()`), which also runs concurrently.

**NOTE**: step 2 is blocked by step 1, which means we can only get the first outer row once we have finished building the hash table based on the fetched inner rows. 

In fact we can start the inner and outer workers together, in this PR, the execution flow of `fetchOuterChunks()` is changed to:
- fetch the first batch of outer rows (stored in a `Chunk`) from the outer child
- blocked until we have finished building hash table based on the inner rows. **NOTE**: this wait and check operation should only take once.
- dispatch the fetched outer rows to the join workers, and continue fetching and dispatching outer rows.

## What are the type of the changes (mandatory)?

- Improvement (non-breaking change which is an improvement to an existing feature)

## How has this PR been tested (mandatory)?

- unit test
- explain test

## Does this PR affect documentation (docs/docs-cn) update? (optional)

No

## Benchmark result if necessary (optional)

benchmarks will be provided later

### CODE REVIEW

**Comment by shenli** [comment]
Good job!
Could compare this branch and master with TPC-H Q17?

**Comment by zz-jason** [comment]
@shenli Yes, it's part of the performance test.

**Comment by zz-jason** [comment]
@shenli Yes, it's part of the performance test.

File: executor/join.go [hunk 1]
```diff
@@ -218,13 +221,42 @@ func (e *HashJoinExec) fetchOuterChunks(ctx context.Context) {
...
+		}
+
 		if outerResult.NumRows() == 0 {
 			return
 		}
 		outerResource.dest <- outerResult
 	}
 }
 
+func (e *HashJoinExec) wait4InnerOnce() (err error, finished bool) {
```
Comment (lysu) [inline executor/join.go]: `error` maybe better as last return item in go's idiom~

File: executor/join.go [hunk 2]
```diff
@@ -218,13 +221,42 @@ func (e *HashJoinExec) fetchOuterChunks(ctx context.Context) {
 			}
 			return
 		}
+
+		if !hasWaitedForInner {
```
Comment (winoros) [inline executor/join.go]: Why not do this before the for loop?

Comment (zz-jason) [reply executor/join.go]: actually we should do this after the first function call of `e.outerExec.Next()`

File: executor/join.go (hunk 1 above)
Comment (zz-jason) [reply executor/join.go]: Yes, but I prefer to do it the same way, which use`error` as the first return value, as our existing old code does

File: executor/join.go (hunk 2 above)
Comment (winoros) [reply executor/join.go]: ok, got it.

File: executor/join.go [hunk 3]
```diff
@@ -218,13 +221,42 @@ func (e *HashJoinExec) fetchOuterChunks(ctx context.Context) {
...
+			innerErr, jobFinished := e.wait4InnerOnce()
+			if jobFinished {
```
Comment (shenli) [inline executor/join.go]: Check the error first.

File: executor/join.go [hunk 4]
```diff
@@ -218,13 +221,42 @@ func (e *HashJoinExec) fetchOuterChunks(ctx context.Context) {
...
+				return
+			} else if innerErr != nil {
+				e.joinResultCh <- &hashjoinWorkerResult{
+					err: errors.Trace(err),
```
Comment (shenli) [inline executor/join.go]: Should we return the innerErr?

File: executor/join.go (hunk 1 above)
Comment (shenli) [reply executor/join.go]: The function doesn't guarantee the `once`. So wait4Inner is a more suitable name.

**Pull request merged**